code/hsec-tools/src/Security/Advisories/Format.hs

```diff
@@ -0,0 +1,432 @@
+codecOS = Toml._TextBy toStr fromStr
+  where
+    toStr =
+      \case
+        MacOS -> "darwin"
+        FreeBSD -> "freebsd"
+        Linux -> "linux"
+        Android -> "linux-android"
+        Windows -> "mingw32"
+        NetBSD -> "netbsd"
+        OpenBSD -> "openbsd"
+    fromStr =
+      \case
+        "darwin" -> pure MacOS
+        "freebsd" -> pure FreeBSD
+        "linux" -> pure Linux
+        "linux-android" -> pure Android
+        "mingw32" -> pure Windows
+        "netbsd" -> pure NetBSD
+        "openbsd" -> pure OpenBSD
+        other -> Left $ "Invalid OS: " <> T.pack (show other)
+
+codecReferenceType :: Toml.Key -> Toml.TomlCodec ReferenceType
+codecReferenceType = Toml.textBy toStr fromStr
+  where
+    toStr x =
+      case lookup x referenceTypes of
+        Just a -> a
+        Nothing -> error $ "Cannot render reference.type " <> show x
+    fromStr x =
+      case lookup x (swap <$> referenceTypes) of
+        Just a -> pure a
+        Nothing -> Left $ "'" <> x <> "' reference.type should be one of: " <> T.intercalate ", " (snd <$> referenceTypes)
+
+defaultingArrayOf :: Toml.TomlBiMap a Toml.AnyValue -> Toml.Key -> Toml.TomlCodec [a]
+defaultingArrayOf d = Toml.dimap Just (fromMaybe mempty) . Toml.dioptional . Toml.arrayOf d
+
+mergeOob ::
+  AttributeOverridePolicy ->
+  -- | out-of-band value
+  Maybe a ->
+  -- | key
+  T.Text ->
+  -- | in-band-value
+  Maybe a ->
+  -- | when key and out-of-band value absent
+  b ->
+  -- | when value present
+  (a -> b) ->
+  -- | when an error occurs
+  (T.Text -> b) ->
+  b
+mergeOob policy oob k ib absent present sendError = do
+  case (oob, ib) of
+    (Just l, Just r) -> case policy of
+      NoOverrides -> sendError $ "illegal out of band override: " <> k
+      PreferOutOfBand -> present l
+      PreferInBand -> present r
+    (Just a, Nothing) -> present a
+    (Nothing, Just a) -> present a
+    (Nothing, Nothing) -> absent
+
+mergeOobOptional ::
+  Applicative f =>
+  AttributeOverridePolicy ->
+  -- | out-of-band value
+  Maybe a ->
+  -- | key
+  T.Text ->
+  -- | in-band-value
+  Maybe a ->
+  -- | when an error occurs
+  (T.Text -> f (Maybe a)) ->
+  f (Maybe a)
+mergeOobOptional policy oob k ib =
+  mergeOob policy oob k ib (pure Nothing) (pure . Just)
+
+mergeOobMandatory ::
+  Applicative f =>
+  AttributeOverridePolicy ->
+  -- | out-of-band value
+  Maybe a ->
+  -- | key
+  T.Text ->
+  -- | in-band value
+  Maybe a ->
+  -- | when an error occurs
+  (T.Text -> f a) ->
+  f a
+mergeOobMandatory policy oob k ib sendError =
+  mergeOob policy oob k ib (sendError $ "missing mandatory key: " <> k) pure sendError
+
+-- | A solution to an awkward problem: how to delete the TOML
+-- block.  We parse into this type to get the source range of
+-- the first block element.  We can use it to delete the lines
+-- from the input.
+newtype FirstSourceRange = FirstSourceRange (First SourceRange)
+  deriving (Show, Semigroup, Monoid)
```

Comment:
I think we should use `DerivingStrategies`
```suggestion
newtype FirstSourceRange = FirstSourceRange (First SourceRange)
  deriving stock Show
  newtype (Semigroup, Monoid)
```

Reply:
*actually*, we might also use `DerivingVia` and don't wrap `First` *at all*, if we only use it for `Monoid` and `Semigroup` instances.
```suggestion
newtype FirstSourceRange = FirstSourceRange SourceRange
  deriving stock Show
  deriving (Semigroup, Monoid) via (First SourceRange)
```